File: Exams/Resoluções de Exames by Carlos Martins/synchs-optimized/Semaphore_.cs
/**
 *
 *  ISEL, LEIC, Concurrent Programming
 *
 *  Semaphore with fast-path optimization
 * 
 *  Carlos Martins, May 2018
 *
 **/

using System;
using System.Threading;

public sealed class Semaphore_ {

	private volatile int permits;
	private volatile int waiters;
	private readonly object monitor;
	 
	// Constructor
	public Semaphore_(int initial = 0) {
		if (initial < 0)
			throw new ArgumentException();
		monitor = new object();
		permits = initial;
	}

	// tries to acquire one permit
	private bool TryAcquire() {
		int p;
		do {
			if ((p = permits) == 0)
				return false;
		} while (Interlocked.CompareExchange(ref permits, p - 1, p) != p);
		return true;
	}
	
	// releases one permit
	private void DoRelease() {
		Interlocked.Increment(ref permits);	// inserts a full memory fence
	}
	
	// Acquire one permit from the semaphore
	public bool Acquire(int timeout = Timeout.Infinite) {
		// try to acquire one permit, if available
		if (TryAcquire())
			return true;
		
		// no permits available; if a null time out was specified, return failure.
		if (timeout == 0)
			return false;

		// if a time out was specified, get a time reference
		TimeoutHolder th = new TimeoutHolder(timeout);
		
		lock(monitor) {
			
			// the current thread declares itself as a waiter..
			waiters++;
			/**
			 * .NET 2.0 Memory Model does not guarantees non-reordering of previous volatile write
			 *  of "waiters" with the next volatile read of "permits", so we must EXPLICITLY
			 *  insert a Full Fence to ensure that this algorithm works!
			 */
			Interlocked.MemoryBarrier();
			try {		
				do {
					// after increment waiters, we must recheck if acquire is possible!
					if (TryAcquire())
						return true;
					// check if the specified timeout expired
					if ((timeout = th.Value) == 0)
						return false;
					Monitor.Wait(monitor, timeout);
				} while (true);
			} catch (ThreadInterruptedException) {
				// if we were interrupted and there are permits available, we can have
				// been notified and interrupted.
				// so, we leave this method throwing ThreadInterruptException, but before
				// we regenerate the notification, if there are available permits
				//
				if (permits > 0) 
					Monitor.Pulse(monitor);
				throw; // re-throw thread interrupted exception
			} finally {
				// the current thread is no longer a waiter
				waiters--;
			}	
		}
	}
		
	// Release one permit
	public void Release(){
		DoRelease();			// inserts a full memory fence
		if (waiters > 0) {	
			lock(monitor) {
				// we must recheck "waiters" after acquire the lock,
				// to avoid unnecessary notifications 
				if (waiters > 0)
					Monitor.Pulse(monitor); // only one thread can proceed execution
			}
		}
	}

	/*
     * Test code.
	 */
	
	public static bool IsCurrentThreadInterrupted() {
		try {
			Thread.Sleep(0);
			return false;
		} catch (ThreadInterruptedException) {
			// re-assert interruption
			Thread.CurrentThread.Interrupt();
			return true;
		}
	}
					
	private static bool TestSemaphoreAsLock() {

		const int MIN_ACQUIRE_TIMEOUT = 5;
		const int MAX_ACQUIRE_TIMEOUT = 50;
		const int MAX_CRITICAL_SECTION_TIME = 5;
		const int JOIN_TIMEOUT = 50;		
		const int RUN_TIME = 10 * 1000;
		const int THREADS = 10;

		Thread[] tthrs = new Thread[THREADS];
		int[] privateCounters = new int[THREADS];
		int[] timeouts = new int[THREADS];
		int sharedCounter = 0;
		Semaphore_ lockSem = new Semaphore_(1);
		
		//
		// Create and start acquirer/releaser threads
		//
		
		for (int i = 0; i < THREADS; i++) {
			int tid = i;
			tthrs[i] = new Thread(() => {
				Random rnd = new Random(tid);
				Console.WriteLine($"-> #{tid:D2} starting...");			
				do {
					try {
						do {
							if (lockSem.Acquire(rnd.Next(MIN_ACQUIRE_TIMEOUT, MAX_ACQUIRE_TIMEOUT)))
								break;
							if (++timeouts[tid] % 100 == 0)
								Console.Write('.');
						} while (true);
						try {
							Interlocked.Increment(ref sharedCounter);
							if (++privateCounters[tid] % 10 == 0) {
								Console.Write($"[#{tid:D2}]");
							}
							Thread.Sleep(rnd.Next(MAX_CRITICAL_SECTION_TIME));
						} finally {
							lockSem.Release();							
						}
					} catch (ThreadInterruptedException) {
						/*
						if (tid == 0)
							do {} while (true);
						*/
						goto exit;
					}
				} while (!IsCurrentThreadInterrupted());
			exit:
				Console.WriteLine($"<- #{tid:00} exiting...");
			});
			tthrs[i].IsBackground = true;
			tthrs[i].Start();
		}

		// run the test threads for a while...
		Thread.Sleep(RUN_TIME);
		
		// Interrupt each test thread and wait for a while until it finished.
		int stillRunning = 0;
		for (int i = 0; i < THREADS; i++) {
			tthrs[i].Interrupt();
			if (!tthrs[i].Join(JOIN_TIMEOUT))
				stillRunning++;
		}
		
		if (stillRunning > 0) {
			Console.WriteLine("\n*** failure: {0} test thread(s) did not answer to interruption", stillRunning);
			return false;
		}
		
		// All threads finished - compute results
		
		Console.WriteLine("\nPrivate counters:\n");
		int sum = 0;
		for (int i = 0; i < THREADS; i++) {
			sum += privateCounters[i];
			if ((i % 5) == 0)
				Console.WriteLine();
			else
				Console.Write(' ');
			Console.Write($"[#{i:00}:{privateCounters[i],4}/{timeouts[i]}]");
		}
		Console.WriteLine($"\n--shared aquisitions: {sharedCounter}, private acquisitions: {sum}");
		return sum == sharedCounter;
	}

	private static bool TestSemaphoreInAProducerConsumerContext() {

		const int MAX_PRODUCE_TIME = 2;
		const int MAX_CONSUME_TIME = 3;
		const int RUN_TIME = 10 * 1000;
		const int JOIN_TIMEOUT = 50;
		const int PRODUCER_THREADS = 5;
		const int CONSUMER_THREADS = 10;
 
		Thread[] cthrs = new Thread[CONSUMER_THREADS];
		Thread[] pthrs = new Thread[PRODUCER_THREADS];
		int[] consumerCounters = new int[CONSUMER_THREADS];
		int[] producerCounters = new int[PRODUCER_THREADS];
		
		// Using our semaphore...
		Semaphore_ freeSem = new Semaphore_(1);
		Semaphore_ dataSem = new Semaphore_(0);
		
		// Create and start consumer threads.		
		for (int i = 0; i < CONSUMER_THREADS; i++) {
			int tid = i;
			cthrs[i] = new Thread(() => {
				Random rnd = new Random(tid);
				do {
					try {
						dataSem.Acquire();
						try {
							if (++consumerCounters[tid] % 10 == 0)
								Console.Write($"[#c{tid:D2}]");
							Thread.Sleep(rnd.Next(MAX_CONSUME_TIME));
						} finally {
							freeSem.Release();
						}
					} catch (ThreadInterruptedException) {
						break;
					}
				} while (!IsCurrentThreadInterrupted());					
			});
			cthrs[i].IsBackground = true;
			cthrs[i].Priority = ThreadPriority.Highest;
			cthrs[i].Start();
		}
		
		// Create and start producer threads.
		for (int i = 0; i < PRODUCER_THREADS; i++) {
			int tid = i;
			pthrs[i] = new Thread(() => {				
				Random rnd = new Random(tid + CONSUMER_THREADS);
				do {
					try {
						freeSem.Acquire();
						try {
							if (++producerCounters[tid] % 10 == 0)
								Console.Write($"[#p{tid:D2}]");
							Thread.Sleep(rnd.Next(MAX_PRODUCE_TIME));	
						} finally {
							dataSem.Release();							
						}
					} catch (ThreadInterruptedException) {
						break;
					}
				} while (!IsCurrentThreadInterrupted());
			});
			pthrs[i].IsBackground = true;
			pthrs[i].Start();
		}
		
		// run the test for a while
		Thread.Sleep(RUN_TIME);
		
		// Interrupt each consumer thread and wait for a while until it finished.
		int stillRunning = 0;
		for (int i = 0; i < CONSUMER_THREADS; i++) {
			cthrs[i].Interrupt();
			if (!cthrs[i].Join(JOIN_TIMEOUT))
				stillRunning++;
		}

		// Interrupt each producer thread and wait for a while until it finished.
		for (int i = 0; i < PRODUCER_THREADS; i++) {
			pthrs[i].Interrupt();
			if (!pthrs[i].Join(JOIN_TIMEOUT))
				stillRunning++;
		}
		
		if (stillRunning > 0) {
			Console.WriteLine($"\n*** failure: {stillRunning} test thread(s) did not answer to interruption");
			return false;
		}
		
		
		// Compute results
		
		Console.WriteLine("\nConsumer counters:");
		int consumptions = 0;
		for (int i = 0; i < CONSUMER_THREADS; i++) {
			consumptions += consumerCounters[i];
			if (i != 0) {
				if (i % 4 == 0)
					Console.WriteLine();
				else
					Console.Write(' ');
			}
			Console.Write($"#c[{i:00}: {consumerCounters[i],4}]");
		}
		if (dataSem.TryAcquire())
			consumptions++;
		
		Console.WriteLine("\nProducer counters:");
		int productions = 0;
		for (int i = 0; i < PRODUCER_THREADS; i++) {
			productions += producerCounters[i];
			if (i != 0) {
				if (i % 4 == 0)
					Console.WriteLine();
				else
					Console.Write(' ');
			}
			Console.Write($"[#p{i:00}: {producerCounters[i],4}]");
		}
		Console.WriteLine($"\n--productions: {productions}, consumptions: {consumptions}");
		return consumptions == productions;
	}
	
	public static void Main() {
	
		Console.WriteLine("\n-->Test semaphore as lock: {0}",
						  TestSemaphoreAsLock() ? "passed" : "failed");
		
		Console.WriteLine("\n-->Test semaphore in a producer/consumer context: {0}",
						  TestSemaphoreInAProducerConsumerContext() ? "passed" : "failed");
	
	}
}

File: Exams/Resoluções de Exames by Carlos Martins/synchs-optimized/Semaphore_.java
/**
 *
 *  ISEL, LEIC, Concurrent Programming
 *
 *  Semaphore with fast-path optimization
 * 
 *  Carlos Martins, May 2018
 *
 **/

import java.util.Random;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;
import java.util.concurrent.locks.*;

public final class Semaphore_ {

	private final AtomicInteger permits;
	private volatile int waiters;
	private final Lock lock;
	private final Condition permitsAvailable;
	 
	// Constructor
	public Semaphore_(int initial) {
		if (initial < 0)
			throw new IllegalArgumentException();
		lock = new ReentrantLock();
		permitsAvailable = lock.newCondition();
		permits = new AtomicInteger(initial);
	}
	
	public Semaphore_() { this(0); }
	
	// tries to acquire one permit
	public boolean tryAcquire() {
		int p;
		do {
			if ((p = permits.get()) == 0)
				return false;
		} while (!permits.compareAndSet(p, p - 1));
		return true;
	}
	
	// releases one permit
	private void doRelease() {
		permits.incrementAndGet();
		// Java guarantees that this write is visible before any subsequent reads
	}
	
	// Acquire one permit from the semaphore
	public boolean acquire(long timeout, TimeUnit unit) throws InterruptedException {
		// try to acquire one permit, if available
		if (tryAcquire())
			return true;
		
		// no permits available; if a null time out was specified, return failure.
		if (timeout == 0)
			return false;

		// if a time out was specified, get a time reference
		boolean timed = timeout > 0;
		long nanosTimeout = timed ? unit.toNanos(timeout) : 0L;
		
		lock.lock();
		try {
			
			// the current thread declares itself as a waiter..
			waiters++;
			/**
			 * Java: JMM guarantees non-ordering of previous volatile write of "waiters"
			 * with the next volatile read of "permits"
			 */
			try {		
				do {
					// after increment waiters, we must recheck if acquire is possible!
					if (tryAcquire())
						return true;
					// check if the specified timeout expired
					if (timed && nanosTimeout <= 0)
						return false;
					if (timed)
						nanosTimeout = permitsAvailable.awaitNanos(nanosTimeout);
					else
						permitsAvailable.await();
				} while (true);
			/**
			 * NOTE: Although this was true in the early versions of Java, it was possible to lose
			 *	  notifications due to the interruption of blocked threads, in version 1.8 of Java
			 *	  it is verified that this behavior is no longer verified.
			 *	  Therefore, it is no longer necessary to capture the interrupt exception for the
			 *	  purpose of regenerating a lost notification.
			 */
			/**
			} catch (InterruptedException ie) {
				// if we were interrupted and there are permits available, we can have
				// been notified and interrupted.
				// so, we leave this method throwing ThreadInterruptException, but before
				// we regenerate the notification, if there are available permits
				//
				if (permits.get() > 0) 
					permitsAvailable.signal();
				throw ie; // re-throw thread interrupted exception
			*/	
			} finally {
				// the current thread is no longer a waiter
				waiters--;
			}	
		} finally {
			lock.unlock();
		}
	}
	
	public void acquire() throws InterruptedException {
		acquire(-1, TimeUnit.MILLISECONDS);
	}

	public boolean acquire(int timeoutMillis) throws InterruptedException {
		return acquire(timeoutMillis, TimeUnit.MILLISECONDS);
	}
	
	// Release one permit
	public void release(){
		doRelease();
		if (waiters > 0) {	
			lock.lock();
			try  {
				if (waiters > 0)
					permitsAvailable.signal(); // only one thread can proceed execution
			} finally {
					lock.unlock();
			}
		}
	}

	/*
     * Test code.
	 */
			
	private static boolean testSemaphoreAsLock() throws InterruptedException {

		final int MIN_ACQUIRE_TIMEOUT = 5;
		final int MAX_ACQUIRE_TIMEOUT = 50;
		final int MIN_CRITICAL_SECTION_TIME = 0;
		final int MAX_CRITICAL_SECTION_TIME = 5;
		final int JOIN_TIMEOUT = 50;		
		final int RUN_TIME = 10 * 1000;
		final int THREADS = 10;

		Thread[] tthrs = new Thread[THREADS];
		int[] privateCounters = new int[THREADS];
		int[] timeouts = new int[THREADS];
		final AtomicInteger sharedCounter = new AtomicInteger();
		Semaphore_ lockSem = new Semaphore_(1);
		
		//
		// Create and start acquirer/releaser threads
		//
		
		for (int i = 0; i < THREADS; i++) {
			int tid = i;
			tthrs[i] = new Thread(() -> {
				Random rnd = new Random(tid);
				System.out.printf("-> #%02d starting...%n", tid);			
				outerLoop: do {
					try {
						do {
							if (lockSem.acquire(rnd.nextInt(MAX_ACQUIRE_TIMEOUT) + MIN_ACQUIRE_TIMEOUT, TimeUnit.MILLISECONDS)) {
								break;
							}
							if (++timeouts[tid] % 1000 == 0)
								System.out.print('.');
						} while (true);
						try {
							sharedCounter.incrementAndGet();
							if (++privateCounters[tid] % 100 == 0) {
								System.out.printf("[#%d]", tid);
							}
							Thread.sleep(rnd.nextInt(MAX_CRITICAL_SECTION_TIME) + MIN_CRITICAL_SECTION_TIME);
						} finally {
							lockSem.release();							
						}
					} catch (InterruptedException ie) {
						/*
						if (tid == 0)
							do {} while (true);
						*/
						break outerLoop;
					}
				} while (!Thread.currentThread().isInterrupted());
				System.out.printf("<- #%02d exiting...%n", tid);
			});
			tthrs[i].setDaemon(true);
			tthrs[i].start();
		}

		// run the test threads for a while...
		Thread.sleep(RUN_TIME);
		
		// Interrupt each test thread and wait for a while until it finished.
		int stillRunning = 0;
		for (int i = 0; i < THREADS; i++) {
			tthrs[i].interrupt();
			tthrs[i].join(JOIN_TIMEOUT);
			if (tthrs[i].isAlive())
				stillRunning++;
		}
		
		if (stillRunning > 0) {
			System.out.printf("%n*** failure: %d test thread(s) did not answer to interruption%n", stillRunning);
			return false;
		}
		
		// All thread finished - compute results
		
		System.out.printf("%nPrivate counters:%n");
		int sum = 0;
		for (int i = 0; i < THREADS; i++) {
			sum += privateCounters[i];
			if (i != 0) {
				if ((i % 4) == 0)
					System.out.println();
				else
					System.out.print(' ');
			}
			System.out.printf("[#%02d: %4d/%d]", i, privateCounters[i], timeouts[i]);
		}
		System.out.printf("%n--shared aquisition: %d, private acquisitions: %d%n", sharedCounter.get(), sum);
		return sum == sharedCounter.get();
	}

	private static boolean testSemaphoreInAProducerConsumerContext() throws InterruptedException {

		final int MAX_PRODUCE_TIME = 5;
		final int MAX_CONSUME_TIME = 5;
		final int RUN_TIME = 10 * 1000;
		final int JOIN_TIMEOUT = 50;
		final int PRODUCER_THREADS = 20;
		final int CONSUMER_THREADS = 30;
 
		Thread[] cthrs = new Thread[CONSUMER_THREADS];
		Thread[] pthrs = new Thread[PRODUCER_THREADS];
		int[] consumerCounters = new int[CONSUMER_THREADS];
		int[] producerCounters = new int[PRODUCER_THREADS];
		
		// Using our semaphore...
		Semaphore_ freeSem = new Semaphore_(1);
		Semaphore_ dataSem = new Semaphore_(0);
		
		// or using the Java semaphore...
		//Semaphore freeSem = new Semaphore(1);
		//Semaphore dataSem = new Semaphore(0);
			
			
		// Create and start consumer threads.
		
		for (int i = 0; i < CONSUMER_THREADS; i++) {
			final int tid = i;
			cthrs[i] = new Thread(() -> {
				Random rnd = new Random(tid);
				do {
					try {
						dataSem.acquire();
						try {
							if (++consumerCounters[tid] % 20 == 0)
								System.out.printf("[#c%d]", tid);
							Thread.sleep(rnd.nextInt(MAX_CONSUME_TIME));
						} finally {
							freeSem.release();
						}
					} catch (InterruptedException ie) {
						break;
					}
				} while (!Thread.currentThread().isInterrupted());					
			});
			cthrs[i].setDaemon(true);
			cthrs[i].setPriority(Thread.MAX_PRIORITY);
			cthrs[i].start();
		}
		
		// Create and start producer threads.
		for (int i = 0; i < PRODUCER_THREADS; i++) {
			final int tid = i;
			pthrs[i] = new Thread(() -> {				
				Random rnd = new Random(tid);
				do {
					try {
						freeSem.acquire();
						try {
							if (++producerCounters[tid] % 20 == 0)
								System.out.printf("[#p%d]", tid);
							Thread.sleep(rnd.nextInt(MAX_PRODUCE_TIME));	
						} finally {
							dataSem.release();							
						}
					} catch (InterruptedException ie) {
						break;
					}
				} while (!Thread.currentThread().isInterrupted());
			});
			pthrs[i].setDaemon(true);
			pthrs[i].start();
		}
		
		// run the test for a while
		Thread.sleep(RUN_TIME);
		
		// Interrupt each consumer thread and wait for a while until it finished.
		int stillRunning = 0;
		for (int i = 0; i < CONSUMER_THREADS; i++) {
			cthrs[i].interrupt();
			cthrs[i].join(JOIN_TIMEOUT);
			if (cthrs[i].isAlive())
				stillRunning++;
		}

		// Interrupt each producer thread and wait for a while until it finished.
		for (int i = 0; i < PRODUCER_THREADS; i++) {
			pthrs[i].interrupt();
			pthrs[i].join(JOIN_TIMEOUT);
			if (pthrs[i].isAlive())
				stillRunning++;
		}
		
		if (stillRunning > 0) {
			System.out.printf("%n*** failure: %d test thread(s) did not answer to interruption%n", stillRunning);
			return false;
		}	
		
		// Compute results
		
		System.out.printf("%nConsumer counters:%n");
		int consumptions = 0;
		for (int i = 0; i < CONSUMER_THREADS; i++) {
			consumptions += consumerCounters[i];
			if (i != 0) {
				if (i % 5 == 0)
					System.out.println();
				else
					System.out.print(' ');
			}
			System.out.printf("[#c%02d: %4d]", i, consumerCounters[i]);
		}
		if (dataSem.tryAcquire())
			consumptions++;
		
		System.out.printf("%nProducer counters:%n");
		int productions = 0;
		for (int i = 0; i < PRODUCER_THREADS; i++) {
			productions += producerCounters[i];
			if (i != 0) {
				if (i % 5 == 0)
					System.out.println();
				else
					System.out.print(' ');
			}
			System.out.printf("[#p%02d: %4d]", i, producerCounters[i]);
		}
		System.out.printf("%n--productions: %d, consumptions: %d%n", productions, consumptions);
		return consumptions == productions;
	}
	
	public static void main(String[] args) throws InterruptedException {
		
		System.out.printf("%n-->Test semaphore as lock: %s%n",
						  testSemaphoreAsLock() ? "passed" : "failed");
		/*
		System.out.printf("%n-->Test semaphore in a producer/consumer context: %s%n",
						  testSemaphoreInAProducerConsumerContext() ? "passed" : "failed");
		*/
		
	}
}
